File: depth-first-search/depth_first_search.ts
function dfs(graph: {[key: number]: number[]}, start: number, visited: Set<number>): void {
    visited.add(start);
    process.stdout.write(start + " ");
    
    const neighbors = graph[start] || [];
    for (const neighbor of neighbors) {
        if (!visited.has(neighbor)) {
            dfs(graph, neighbor, visited);
        }
    }
}

const graph = {
    0: [1, 2],
    1: [2],
    2: [0, 3],
    3: [3]
};

const visited = new Set<number>();
dfs(graph, 2, visited);

File: depth-first-search/depth_first_search.rs
use std::collections::{HashMap, HashSet};

fn dfs(graph: &HashMap<i32, Vec<i32>>, start: i32, visited: &mut HashSet<i32>) {
    visited.insert(start);
    print!("{} ", start);
    
    if let Some(neighbors) = graph.get(&start) {
        for &neighbor in neighbors {
            if !visited.contains(&neighbor) {
                dfs(graph, neighbor, visited);
            }
        }
    }
}

fn main() {
    let mut graph = HashMap::new();
    graph.insert(0, vec![1, 2]);
    graph.insert(1, vec![2]);
    graph.insert(2, vec![0, 3]);
    graph.insert(3, vec![3]);
    
    let mut visited = HashSet::new();
    dfs(&graph, 2, &mut visited);
}